apps/segment/src/modules/db/segment-apl-repository.ts

```diff
@@ -0,0 +1,134 @@
+import { AuthData } from "@saleor/app-sdk/APL";
+import { DeleteItemCommand, GetItemCommand, PutItemCommand, ScanCommand } from "dynamodb-toolbox";
+import { err, ok, ResultAsync } from "neverthrow";
+
+import { BaseError } from "@/errors";
+import { createLogger } from "@/logger";
+import { SegmentAPLEntityType, SegmentMainTable } from "@/modules/db/segment-main-table";
+
+import { SegmentAPLMapper } from "./segment-apl-mapper";
+
+export class SegmentAPLRepository {
+  private logger = createLogger("DynamoDBAPLRepository");
+
+  private segmentAPLMapper = new SegmentAPLMapper();
+
+  static ReadEntityError = BaseError.subclass("ReadEntityError");
+  static WriteEntityError = BaseError.subclass("WriteEntityError");
+  static DeleteEntityError = BaseError.subclass("DeleteEntityError");
+  static ScanEntityError = BaseError.subclass("ScanEntityError");
+
+  constructor(
+    private deps: {
+      segmentAPLEntity: SegmentAPLEntityType;
+    },
+  ) {}
+
+  async getEntry(args: { saleorApiUrl: string }) {
+    const getEntryResult = await ResultAsync.fromPromise(
+      this.deps.segmentAPLEntity
+        .build(GetItemCommand)
+        .key({
+          PK: SegmentMainTable.getAPLPrimaryKey({
+            saleorApiUrl: args.saleorApiUrl,
+          }),
+          SK: SegmentMainTable.getAPLSortKey(),
+        })
+        .send(),
+      (error) =>
+        new SegmentAPLRepository.ReadEntityError("Failed to read APL entity", { cause: error }),
+    );
+
+    if (getEntryResult.isErr()) {
+      this.logger.error("Error while reading APL entity from DynamoDB", {
+        error: getEntryResult.error,
+      });
+
+      return err(getEntryResult.error);
+    }
+
+    if (!getEntryResult.value.Item) {
+      this.logger.warn("APL entry not found", { args });
+
+      return err(new SegmentAPLRepository.ReadEntityError("APL entry not found"));
+    }
+
+    return ok(this.segmentAPLMapper.dynamoDBEntityToAuthData(getEntryResult.value.Item));
+  }
+
+  async setEntry(args: { authData: AuthData }) {
+    const setEntryResult = await ResultAsync.fromPromise(
+      this.deps.segmentAPLEntity
+        .build(PutItemCommand)
+        .item(this.segmentAPLMapper.authDataToDynamoPutEntity(args.authData))
+        .send(),
+      (error) =>
+        new SegmentAPLRepository.WriteEntityError("Failed to write APL entity", { cause: error }),
+    );
+
+    if (setEntryResult.isErr()) {
+      this.logger.error("Error while putting APL into DynamoDB", {
+        error: setEntryResult.error,
+      });
+
+      return err(setEntryResult.error);
+    }
+
+    return ok(undefined);
```

Comment:
do we have anything valuable returned from dynamo?

Reply:
I checked and we have:
* `ConsumedCapacity`
* `ToolboxItem` - entry itself

I'm wondering if we need this data in APL class?

Reply:
I think this happens so rarely it doesn't matter, but preferably if we don't return anything, we can save transfer from Dynamo too (don't fetch what we don't need).